File: MovieCatalogueV2/app/src/main/java/com/google/fdp/moviecataloguev2/repositories/MovieRepository.java
package com.google.fdp.moviecataloguev2.repositories;

import android.app.Application;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import com.google.fdp.moviecataloguev2.BuildConfig;
import com.google.fdp.moviecataloguev2.databases.FavouriteMovie;
import com.google.fdp.moviecataloguev2.databases.FavouriteMovieDao;
import com.google.fdp.moviecataloguev2.databases.FavouriteTvShow;
import com.google.fdp.moviecataloguev2.databases.FavouriteTvShowDao;
import com.google.fdp.moviecataloguev2.databases.RoomMovieDatabase;
import com.google.fdp.moviecataloguev2.fragments.MovieFragment;
import com.google.fdp.moviecataloguev2.models.Movie;
import com.google.fdp.moviecataloguev2.models.responses.BaseResponse;
import com.google.fdp.moviecataloguev2.networks.MovieService;
import com.google.fdp.moviecataloguev2.networks.RetrofitClient;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Created by gama on 2019-07-26.
 * Addin Gama Bertaqwa
 * addingama@gmail.com
 */
public class MovieRepository {
    private FavouriteMovieDao favouriteMovieDao;
    private FavouriteTvShowDao favouriteTvShowDao;
    private MutableLiveData<List<Movie>> mutableLiveData;
    private MovieService service;
    private List<Movie> defaultResponse = new ArrayList<Movie>();

    private static MovieRepository INSTANCE;


    public MovieRepository(Application application) {
        favouriteMovieDao = RoomMovieDatabase.getAppDatabase(application).favouriteMovieDao();
        favouriteTvShowDao = RoomMovieDatabase.getAppDatabase(application).favouriteTvShowDao();
        mutableLiveData = new MutableLiveData<List<Movie>>();
        service = new RetrofitClient().createService(MovieService.class);
    }

    public static  synchronized  MovieRepository getInstance(Application application) {
        if (INSTANCE == null) {
            INSTANCE = new MovieRepository(application);
        }

        return INSTANCE;
    }



    public LiveData<List<Movie>> getAllMovies() {
        return mutableLiveData;
    }

    public void updateData(String type) {
        if (MovieFragment.MOVIE_KEY.equals(type)) {
            service.getMovies(BuildConfig.API_KEY).enqueue(moviesCallback);
        } else {
            service.getTvSeries(BuildConfig.API_KEY).enqueue(moviesCallback);
        }
    }

    public void searchData(String type, String keyword) {
        if (MovieFragment.MOVIE_KEY.equals(type)) {
            service.searchMovies(BuildConfig.API_KEY, keyword).enqueue(moviesCallback);
        } else {
            service.searchTvSeries(BuildConfig.API_KEY, keyword).enqueue(moviesCallback);
        }
    }

    private Callback<BaseResponse<Movie>> moviesCallback = new Callback<BaseResponse<Movie>>() {
        @Override
        public void onResponse(Call<BaseResponse<Movie>> call, Response<BaseResponse<Movie>> response) {
            if(response.isSuccessful()) {
                if (response.body() != null) {
                    mutableLiveData.postValue(response.body().results);
                }

            }
        }

        @Override
        public void onFailure(Call<BaseResponse<Movie>> call, Throwable t) {
            mutableLiveData.postValue(defaultResponse);
        }
    };

    public LiveData<Boolean> isFavourite(Long id, String type) {
        if (MovieFragment.MOVIE_KEY.equals(type)) {
            return Transformations.map(favouriteMovieDao.findById(id),
                it -> !it.isEmpty()
            );
        } else {
            return Transformations.map(favouriteTvShowDao.findById(id),
                    it -> !it.isEmpty()
            );
        }
    }

    public void insertFavourite(Movie data, String type) {
        if (MovieFragment.MOVIE_KEY.equals(type)) {
            favouriteMovieDao.insert(new FavouriteMovie(
                    data.getId(),
                    data.getTitle(),
                    data.getReleaseDate(),
                    data.getOverview(),
                    data.getOriginalLanguage(),
                    data.getPosterPath(),
                    data.getRating(),
                    data.getBackdropPath()
            ));
        } else {
            favouriteTvShowDao.insert(new FavouriteTvShow(
                    data.getId(),
                    data.getTitle(),
                    data.getReleaseDate(),
                    data.getOverview(),
                    data.getOriginalLanguage(),
                    data.getPosterPath(),
                    data.getRating(),
                    data.getBackdropPath()
            ));
        }
    }

    public void deleteFavourite(Movie data, String type) {
        if (MovieFragment.MOVIE_KEY.equals(type)) {
            favouriteMovieDao.delete(new FavouriteMovie(
                    data.getId(),
                    data.getTitle(),
                    data.getReleaseDate(),
                    data.getOverview(),
                    data.getOriginalLanguage(),
                    data.getPosterPath(),
                    data.getRating(),
                    data.getBackdropPath()
            ));
        } else {
            favouriteTvShowDao.delete(new FavouriteTvShow(
                    data.getId(),
                    data.getTitle(),
                    data.getReleaseDate(),
                    data.getOverview(),
                    data.getOriginalLanguage(),
                    data.getPosterPath(),
                    data.getRating(),
                    data.getBackdropPath()
            ));
        }
    }

    public LiveData<List<Movie>> getAllFavourite(String type) {

        if (MovieFragment.MOVIE_KEY.equals(type)) {

            return Transformations.map(favouriteMovieDao.getAllFavoriteMovies(), it -> {
                List<Movie> movies = new ArrayList<Movie>();
                for (FavouriteMovie favouriteMovie : it) {
                    movies.add(constructFavouriteToMovie(favouriteMovie));
                }
                return movies;
            });
        } else {

            return Transformations.map(favouriteTvShowDao.getAllFavoriteTvShows(), it -> {
                List<Movie> tvSeries = new ArrayList<Movie>();
                for (FavouriteTvShow favouriteTvShow : it) {
                    tvSeries.add(constructFavouriteToMovie(favouriteTvShow));
                }
                return tvSeries;
            });
        }
    }

    private Movie constructFavouriteToMovie(Object favourite) {
        if (favourite instanceof FavouriteMovie) {
            FavouriteMovie data = (FavouriteMovie) favourite;
            return new Movie(
                    data.getId(),
                    data.getTitle(),
                    data.getReleased(),
                    data.getOverview(),
                    data.getLanguage(),
                    data.getPoster(),
                    data.getBackdrop(),
                    data.getRating()
            );
        } else if (favourite instanceof FavouriteTvShow) {
            FavouriteTvShow data = (FavouriteTvShow) favourite;
            return new Movie(
                    data.getId(),
                    data.getTitle(),
                    data.getReleased(),
                    data.getOverview(),
                    data.getLanguage(),
                    data.getPoster(),
                    data.getBackdrop(),
                    data.getRating()
            );
        } else {
            return new Movie();
        }
    }
}

File: MovieCatalog 2/app/src/main/java/com/alfianh/moviecatalog/repository/MovieRepository.kt
package com.alfianh.moviecatalog.repository

import android.app.Application
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import com.alfianh.moviecatalog.BuildConfig.API_KEY
import com.alfianh.moviecatalog.database.*
import com.alfianh.moviecatalog.fragments.MovieFragment
import com.alfianh.moviecatalog.models.Movie
import com.alfianh.moviecatalog.models.response.BaseResponse
import com.alfianh.moviecatalog.network.MovieService
import com.alfianh.moviecatalog.network.RetrofitClient
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MovieRepository constructor(application: Application) {

  private val favoriteMovieDao: FavoriteMovieDao = MovieRoomDatabase.getDatabase(
      application).favoriteMovieDao()
  private val favoriteTvShowDao: FavoriteTvShowDao = MovieRoomDatabase.getDatabase(
      application).favoriteTvShowDao()
  private var mutableLiveData = MutableLiveData<List<Movie>>()
  private var service: MovieService = RetrofitClient.createService(MovieService::class.java)

  companion object {
    private var INSTANCE: MovieRepository? = null
    fun getInstance(application: Application): MovieRepository {
      return INSTANCE ?: synchronized(MovieRepository::class) {
        val instance = MovieRepository(application)
        INSTANCE = instance
        instance
      }
    }
  }

  fun getAllMovies(): LiveData<List<Movie>> {
    return mutableLiveData
  }

  fun isFavorite(id: Long, type: String) = Transformations.map(
      if (type == MovieFragment.MOVIE_KEY) {
        favoriteMovieDao.findById(id)
      } else {
        favoriteTvShowDao.findById(id)
      }) {
    it.isNotEmpty()
  }

  fun updateData(type: String) {
    if (type == MovieFragment.MOVIE_KEY) {
      service.getMovies(API_KEY)
    } else {
      service.getTvShows(API_KEY)
    }.enqueue(object : Callback<BaseResponse<Movie>> {
      override fun onFailure(call: Call<BaseResponse<Movie>>, t: Throwable) {
        mutableLiveData.postValue(null)
      }

      override fun onResponse(call: Call<BaseResponse<Movie>>,
          response: Response<BaseResponse<Movie>>) {
        if (response.isSuccessful) {
          mutableLiveData.postValue(response.body()?.results)
        }
      }
    })
  }

  fun getAllFavorite(type: String): LiveData<List<Movie>> {
    if (type == MovieFragment.MOVIE_KEY) {
      return Transformations.map(favoriteMovieDao.getAllFavoriteMovies()) {
        val movies: MutableList<Movie> = mutableListOf()
        it.forEach { favorite ->
          movies.add(constructFavoriteToMovie(favorite))
        }
        movies
      }
    } else {
      return Transformations.map(favoriteTvShowDao.getAllFavoriteTvShows()) {
        val tvShows: MutableList<Movie> = mutableListOf()
        it.forEach { favorite ->
          tvShows.add(constructFavoriteToMovie(favorite))
        }
        tvShows
      }
    }
  }

  suspend fun insertFavorite(favorite: Movie, type: String) {
    if (type == MovieFragment.MOVIE_KEY) {
      favoriteMovieDao.insert(
          FavoriteMovie(id = favorite.id, title = favorite.title, released = favorite.released,
              poster = favorite.poster, language = favorite.language, rating = favorite.rating,
              backdrop = favorite.backdrop, overview = favorite.overview))
    } else {
      favoriteTvShowDao.insert(
          FavoriteTvShow(id = favorite.id, title = favorite.title, released = favorite.released,
              poster = favorite.poster, language = favorite.language, rating = favorite.rating,
              backdrop = favorite.backdrop, overview = favorite.overview))
    }
  }

  suspend fun deleteFavorite(favorite: Movie, type: String) {
    if (type == MovieFragment.MOVIE_KEY) {
      favoriteMovieDao.delete(
          FavoriteMovie(id = favorite.id, title = favorite.title, released = favorite.released,
              poster = favorite.poster, language = favorite.language, rating = favorite.rating,
              backdrop = favorite.backdrop, overview = favorite.overview))
    } else {
      favoriteTvShowDao.delete(
          FavoriteTvShow(id = favorite.id, title = favorite.title, released = favorite.released,
              poster = favorite.poster, language = favorite.language, rating = favorite.rating,
              backdrop = favorite.backdrop, overview = favorite.overview))
    }
  }

  private fun constructFavoriteToMovie(favorite: Any): Movie = when (favorite) {
    is FavoriteMovie -> Movie(id = favorite.id, title = favorite.title,
        released = favorite.released, poster = favorite.poster, language = favorite.language,
        rating = favorite.rating, backdrop = favorite.backdrop, overview = favorite.overview)
    is FavoriteTvShow -> Movie(id = favorite.id, title = favorite.title,
        released = favorite.released, poster = favorite.poster, language = favorite.language,
        rating = favorite.rating, backdrop = favorite.backdrop, overview = favorite.overview)
    else -> Movie()
  }


}